### ISSUE TITLE
MathJax state: \newcommand and friends, numbered eqs

### ISSUE THREAD

**Issue by cben** [cben/CodeMirror-MathJax#2]

**Comment by cben** [comment]
Turns out MathJax has stateful support for \newcommand, and it works.
However it depends on the order they're typeset, which might not correspond to document order
To make it usable:
- [x] Leave `$\newcommand{...}$` in source form instead of becoming invisible (special case of #10).
- [ ] on change to a `$\newcommand...$` formula, re-typeset all (or all following) math.

**Comment by cben** [comment]
- [x] Also pass `\[re]newcommand...` without any `$` etc delimiters to mathjax.
  [Pandoc does this](http://johnmacfarlane.net/pandoc/demo/example9/pandocs-markdown.html#latex-macros).

**Comment by cben** [comment]
Undocumented: `\newcommand` alters `MathJax.InputJax.TeX.Definitions.macros` and `MathJax.InputJax.TeX.Definitions.environment`.  Could try to save/restore them instead of re-typesetting all following math.
However, it's rare enough that re-typesetting should be acceptable.

**Comment by cben** [comment]
Simular (but worse?) issues of rendering order sensitivity happen with `\begin{equation}` numbering.  See https://github.com/mathjax/MathJax/issues/422.

This is going to be more acute if I optimize initial pass (#14) to be partial/out-of-order.
- [ ] Learn which phase of MJ updates state (macros, numbers, labels).  Preprocess?  Process?  Render?

**UPDATE:** MJ 2.7 added `resetEquationNumbers()` API https://docs.mathjax.org/en/v2.7/advanced/typeset.html#reset-automatic-equation-numbering, in MJ 3 it's `texReset()`: https://docs.mathjax.org/en/v3.2/web/typeset.html#resetting-automatic-equation-numbering

**Comment by cben** [comment]
http://jsfiddle.net/89r6p/3/ demonstates mathjax running again (click) on labels, numbers & refs.

**Comment by cben** [comment]
Numbering state is in `MathJax.Extension["TeX/AMSmath"].number` and `MathJax.Extension["TeX/AMSmath"].startNumber`; there is a host of other variables to handle labels&refs: https://github.com/mathjax/MathJax/blob/master/unpacked/extensions/TeX/AMSmath.js#L562

**Comment by cben** [comment]
- [ ] syntax highlight \begin...\end
- [ ] syntax highlight \newcommand...}
- [ ] avoid "label foo multiply defined" which makes labels completely unusable

**Comment by cben** [comment]
- [ ] handle `\DeclareMathOperator` not wrapped in $...$
  - [ ] handle any tex macros not wrapped in math?  Probably WONTFIX.

Should check how pandoc handles those.
